### ISSUE TITLE
Replacing XNA Namespaces

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#3306]
A few years ago we discussed replacing the `Microsoft.Xna.` namespace with `MonoGame.`, but we felt it was too soon as WP7 and Xbox 360 was still pretty active.

So are we getting closer?  Is 2015 the year we make this move?

Can we use some sort of assembly alias as a fallback for older code?

What are issues to consider?  Like how would things like binary deserialization be affected which depend on `ContentTypeReader` names?

**Comment by cra0zy** [comment]
Just so nobody does it, if you want to add "+1" or "-1" don't write it bellow but instead just thumbs up or down the original issue above.

**Comment by InfiniteProductions** [comment]
As no one seems to have point to the most important aspect of this potential renaming, I do it: Microsoft may, or may not, be very happy that some "third party devs" use their names and brands without authorization (if it is the case), so change of namespace's name could be good there (who knows ? US and "lawyers craziness" can lead to anything).

Apart from that, I don't mind IMHO, namespace can be anything as long as features are there (but perhaps removing MS name will also push bugs away :D)

**Comment by KonajuGames** [comment]
> Microsoft may, or may not, be very happy that some "third party devs" use their names and brands without authorization (if it is the case)

Microsoft have been great supporters of the project, and have developed training resources and submitted code to the project.  I wouldn't worry about that.

**Comment by tomspilman** [comment]
@dellis1972 [brought up type forwarding](http://community.monogame.net/t/monogame-feature-wishlist/6850/55?u=tom) as a possible solution:

 https://msdn.microsoft.com/en-us/library/ms404275(v=vs.110).aspx

Not sure if it does, but it is worth investigating.

**Comment by KonajuGames** [comment]
I'm not sure it will fix it either. Type forwarding is per type, so it would need an entry for every public  type in MonoGame. Also, it is not supported in all runtimes.https://msdn.microsoft.com/en-us/library/ms177220(v=vs.110).aspx

**Comment by kevinstowepcs** [comment]
Why not put a compiler directive on the namespace declaration in the monogame libraries or is that not an option? Allow people to simply specify a compiler variable if they need legacy support

**Comment by cra0zy** [comment]
> Why not put a compiler directive on the namespace declaration in the monogame libraries or is that not an option? Allow people to simply specify a compiler variable if they need legacy support

So much work... O o

**Comment by KonajuGames** [comment]
> Why not put a compiler directive on the namespace declaration in the monogame libraries or is that not an option? Allow people to simply specify a compiler variable if they need legacy support

C# compilers do not have preprocessor macro support like C++.  The extent of conditional symbols in C# is if a symbol is defined or not.

**Comment by cra0zy** [comment]
> C# compilers do not have preprocessor macro support like C++. The extent of conditional symbols in C# is if a symbol is defined or not.

Still, his idea is valid, we only need to select between 2 options, Xna namespace and MG namespace.

**Comment by tomspilman** [comment]
> Still, his idea is valid, we only need to select between 2 options

Bigger issue is everything else.

So we then ship both sets of namespaces in the installers for each platform?  Which ones do we hook up to the templates?  Which ones are used by the Pipeline tool?  Do we generate two sets of docs?

I'm starting to think we may just need to make a clean break for new namespaces in the develop branch, then keep the old namespaces in a maintenance mode in another branch.

**Comment by kevinstowepcs** [comment]
That would work, if you need legacy namespaces then checkout the legacy branch. It allows for you to eventually phase it out. We are what? 6 years past XNA's departure or something. Right now, by keeping the XNA namespace around we are further exacerbating the issue because all new games written in monogame are the xna namespace as well. Its like a bandaid just gotta rip it off :) (Easier said than done, i know)

**Comment by Harag9** [comment]
As monogame is NOT XNA anymore and it's been years since XNA departure I think you should start branding monogame and change the namespace - Yes, we would have to change our existing solutions and recompile, but if you do it from v4 onwards that would be a good time to make the change. Go for it!

**Comment by simoneddeland** [comment]
Have there been any further internal discussions about changing the namespaces?

**Comment by KonajuGames** [comment]
There has not been any further discussion on this as we are all busy with projects.  I think aiming for a 4.0 release with new namespaces is the best option so far.  There are a few things to be done first, such as clearing out some of the current PRs, as this change would likely conflict with every PR that hasn't been merged at that time.

**Comment by LonghronShen** [comment]
As a developer of a game engine which is based on XNA and MonoGame, I do love XNA's spirit and I want XNA lives. Since XNA itself is somewhat small minority, the people who loves XNA should be united to keep MonoGame compatible with XNA, that makes existing XNA documents, samples and other resources still usable. In fact, there is no need to rebuild those resources again. Changing namespaces doesn't make much sense and that would make existing code broken. I do agree with that MonoGame should have a new start, but I think we should start from supporting new technologies rather than entangled in namespace.

**Comment by RUSshy** [comment]
What about two repo, MonoGame-Legacy, with namespace unchanged, and only bug-fix updates, and MonoGame with new namespace and features/breaking changes

**Comment by nkast** [comment]
A less aggressive route would be to keep the legacy APIs and move all new APIs/extensions under a new namespace. A similar change can be made to the folders and organize source files under /XNA.Framework/ and /Monogame.Framework (or something similar)

**Comment by willmotil** [comment]
Well i have been thinking about this from time to time. 
When monogame 4.0 hits it should be really close to xna 4.0. 
xna projects should still be easily portable at that time. 
I think that is the practical primary requirement. 

When mg 4.0 switches to 5.0 i think that would probably be the time to yank the xna namespaces it would also be a clear point were you start to say well this is more then xna were 4.0 is really close.

It's not to say monogame 4.0 has to be long lived to switch to 5.0.
I do think you should do a official version release just before a namespace change and it should be 4.0 for clarity to people porting over that in 5.0 is much more and 4.0 is really close to xna 4.0.

That was my thoughts

**Comment by lloydjatkinson** [comment]
2019, pretty weird still seeing XNA in the docs!

**Comment by lloydjatkinson** [comment]
It's been 8 years since this issue was opened. Today I was considering maybe trying out MonoGame again and the docs and namespaces still have XNA all over it.

What is the hold up, exactly? It definitely makes MonoGame feel like it's not really being maintained, especially as the 4.0 release never materialised.

**Comment by MrGrak** [comment]
console support and backwards compatibility would be my guess. replacing xna with mg namespaces seems like a lot of work for little return. plus, there is an alternative perspective to have here: monogame is very stable. code written in 2012 will still more or less work just fine in 2022, and that can be beneficial for those porting older codebases to new platforms, which seems to be one of mg's audiences. there are many povs to see

**Comment by SimonDarksideJ** [comment]
Personally, other than naming, changing the NameSpaces does not actually add any value and only incurs pain, plus it creates issues with the many tutorials, videos and more that have invested time to educate others on the use of MonoGame, it's more than just the API.

MonoGame was always the evolution of XNA and raising the torch to a framework that made building games more fun and easier for coders, so keeping the XNA heritage costs nothing.

My two pence worth.

**Comment by Xrayez** [comment]
As someone who discovered MonoGame just recently, I've found existing namespaces to be off-putting and confusing. Not because of Microsoft as a company specifically, certainly not, but it gives an *impression* that:

- MonoGame itself is incomplete (not sure if this is the case).
- MonoGame is obsolete, deprecated, neglected, and/or may not be actively maintained in the future, because, maybe, just maybe, maintainers believe that MonoGame doesn't have any long-term future in the first place. I'd like to believe otherwise, of course!

Even MonoGame's own documentation dedicates a paragraph to explain why using Microsoft namespaces is required. I think it's definitely worth to replace them with a new major release. It may feel like bikeshedding, but this certainly gives a false sense that the framework itself is neglected by the development team, but even small things like these can potentially undermine the trust of new users. But from what I can tell (given the age of this issue opened), it appears to me that perhaps MonoGame isn't as actively maintained as before.

To summarize, in my subjective opinion, I think that one of the major reasons why those namespaces are still not renamed (excluding technical reasons) is that the very *action* of doing so would promote a particular intention that, perhaps, maintainers are not willing to commit to.

In either case, thanks making this project to exist in the first place! Hopefully my insights inspire action, which may not necessarily manifest as renaming, but via other means, such as regular MonoGame releases! 🙂

**Comment by NicusorN5** [comment]
I agree with the fact that Monogame has an outdated architecture, for example not supporting geometry and tesselation shaders (yet). Also not supporting proper model loading. The model pipeline is the same as in ~2005-ish game engines.

You will have to do great efforts writing boiler plate code if you want to do proper 3D rendering. I'd greatly recommend Stride instead of Monogame for 3D.

Monogame is still very great for writing 2D games. I'd use Monogame anytime over Unity for 2D.

Also the framework is not being negelected by the developers, can you even see the activity in the main repository?

If you didn't notice, the frameworks are the same as in XNA, for backwards compatibility.

**Comment by Xrayez** [comment]
Given reactions above, looks like I need to clarify my message. I invite people who disagree to re-read my previous post. 🙂

I did *not* assert MonoGame being incomplete or neglected. It's all about assumptions and expectations. All I've shown is how the surface of what MonoGame *looks like* from the outside of this community, like having XNA namespaces gives a *false* impression of MonoGame being incomplete or neglected, and it seems to be just one of the aspects that gives this impression. There's a clear difference

You can treat my response as building upon [this comment](https://github.com/MonoGame/MonoGame/issues/3306#issuecomment-515991434) which received a bunch of thumb-ups three years ago here (not by me, obviously). If a project is giving an incorrect expectation on the outside, this may undermine MonoGame's long-term success. This *used to be* my own impression, further reinforced by answers on Q&A forums such as Quora, [like here](https://qr.ae/prvVtw).

Mind you, I'd like MonoGame project to thrive. I'm just giving you the benefit of an outside perspective. Whether you accept it or not, people tend to make decisions based on these surface criteria. If the project does not concern itself with attracting new users, and instead just caters to the userbase with previous XNA experience for the most part, all of the above is certainly not an issue. I have personally not used XNA at all.

Unfortunately, I received a negative response from a MonoGame contributor, who labeled my intentions as "concern trolling". Not at all. I hope you understand that my intentions are good, and I hope that you assume positive intentions too. I'm not just an outsider to this project. I discovered MonoGame a few weeks ago, I even joined Discord server, and summited my first pull request #7974 (hopefully to be merged). Depending on how things go, I may continue contributing to the project!

Therefore, I do *not* think MonoGame is incomplete or neglected. Each project has their own priorities. To speak figuratively, I'm not saying that you don't have a window, I just suggest that, perhaps, the window must be cleaned up.

**Comment by kiddkaffeine** [comment]
> Personally, other than naming, changing the NameSpaces does not actually add any value and only incurs pain, plus it creates issues with the many tutorials, videos and more that have invested time to educate others on the use of MonoGame, it's more than just the API.
> 
> MonoGame was always the evolution of XNA and raising the torch to a framework that made building games more fun and easier for coders, so keeping the XNA heritage costs nothing.
> 
> My two pence worth.

I concur with this opinion.

API stability is a significant benefit, not a negative, because it means it can be trusted for long projects.  Furthermore, when/building third-party libraries as well as multiplatform projects, the API compatibility between MonoGame, FNA, KNI, and other projects has been a major boon.

For new features, I think it's better to build on top of existing APIs.  If we ever enter a situation where major breaking changes have to be made, then we'd probably need some kind of compatibility layer so both the old and new namespaces are available.  But that's even more overhead for MonoGame maintainers.
